### DISCUSSION TITLE
Update Webhooks

### DISCUSSION THREAD

**Discussion by davehakkens** [ONEARMY/community-platform#3238] [Big Updates]
We have a few webhooks installed to make a notification in Discord to the #activitybot channel. 
<img width="777" alt="afbeelding" src="https://github.com/ONEARMY/community-platform/assets/13672737/af2c4ac8-3d08-4204-90ab-dec3ff2f4cbc">

# Current problems

1. They are only configured for Precious Plastic
2. They are only configured for map pin and how-to (not research or questions)
3. They are triggered on edit (should only be triggered on create/approved)
4. They are not documented for developers

# Things to keep in mind
- Keep it simple, since it might not be a long term solution as we plan to move from from Firebase. 

# What is needed:
- [ ] Create an overview and document all the needed webhooks 
- [ ] Double check if the current method. (Firebase functions) is still the best way to implement them
- [ ] Update existing webhooks to not trigger on ‘edit’
- [ ] Add webhooks for missing modules ( New question, research topic update)

**Comment by davehakkens** [comment]
Suggestion for triggers we need

In these events:
- New Update posted in a Research module
- New How-to approved by admin
- New Member Map-pin approved by admin 
- New Space Map-pin approved by admin
- New Question posted in Question module

If any of these events happen.
Send message in Discord channel (#activitybot)

**Reply by davehakkens** [reply]
Formatting to be used:

**Research:**
📝 Nice. [username] posted an update in their research [insert research title] 
Learn about it here: [[url to research update](url)]


**How-to:**
🎁 Yeah! New How-to [insert how-to title] posted by [username] 
Check it out here: [[url to how-to](url)]

**Member-map-pin:**
👋 New pin on the map from [username] that want's to get started
See if they are local to you here:  [[url to map](url)]

**Space-map-pin:**
🌏 New space on the map from [username] that grows the global movement!
Find them here:  [[url to map](url)]

**Question:**
❓ [username] has a new question [insert question title]
Help them out and answer here:  [[url to question](url)]

**Comment by davehakkens** [comment]
Added an issue here: https://github.com/ONEARMY/community-platform/issues/3533

**Comment by goratt12** [comment]
I suggest moving the Discord Webhook logic out of the Firebase Functions themself, like suggested in this PR #3636 .
After that we can start implementing more and more use cases and logic for the Discord Webhook.

**Comment by goratt12** [comment]
We can also implement testing the Discord Webhook by creating a testing channel in discord (with separate webhook) and calling the 
`/webhooks/{webhook.id}/{webhook.token}/messages/{message.id}` endpoint to verify the message was sent.

**Reply by davehakkens** [reply]
Yes. Up for creating that! 
let me know what I should do for that

**Reply by goratt12** [reply]
### Steps to Create Discord Testing Enviroment

1. **Create a Discord Bot**:
   - Go to the [Discord Developer Portal](https://discord.com/developers/applications).
   - Click "New Application" and give it a name.
   - Go to the "Bot" section and click "Add Bot."
   - Click "Reset Token" to generate a bot token. **Copy and save this token** (it will be needed for API calls).

2. **Invite Bot to Server**:
   - Go to the "OAuth2" section of your application in the Developer Portal.
   - Under "OAuth2 URL Generator," select "bot" and "applications.commands."
   - Under "Bot Permissions," select:
     - Read Messages
     - Read Message History
   - Copy the generated URL and provide it to the Discord server admin to invite the bot to the server.

3. **Create a Test Channel in Discord**:
   - In your Discord server, create a new test channel (e.g., `#test-notifications`).

4. **Assign Bot Permissions in the Server**:
   - Ensure the bot has the required permissions in the test channel:
     - Read Messages
     - Read Message History
   - This can be done by going to the channel settings and adjusting the permissions for the bot.

5. **Create a Webhook for the Test Channel**:
   - Go to the settings of the test channel.
   - Under the "Integrations" tab, click "Create Webhook."
   - Provide a name for the webhook (e.g., "Test Notification Webhook").
   - Copy the webhook URL and provide it for integration purposes.

### Details Needed

1. **Bot Token**: The token generated for the bot in the Developer Portal.
2. **Test Channel ID**: The ID of the test channel created for notifications.
3. **Webhook URL**: The URL of the webhook created for the test channel.

**Reply by davehakkens** [reply]
Nice guide! I've send they keys on Discord DM and invited you to the test channel there

**Comment by pizzaisdavid** [comment]
I think this discussion is finished.. right?

**Reply by davehakkens** [reply]
Yep can close indeed, thanks for helping 💪
